apps/emqx/src/emqx_trace/emqx_trace_formatter.erl

```diff
@@ -100,7 +109,11 @@ format_payload(Payload, text) when ?MAX_PAYLOAD_FORMAT_LIMIT(Payload) ->
     unicode:characters_to_list(Payload);
 format_payload(Payload, hex) when ?MAX_PAYLOAD_FORMAT_LIMIT(Payload) -> binary:encode_hex(Payload);
 format_payload(<<Part:?TRUNCATED_PAYLOAD_SIZE/binary, _/binary>> = Payload, Type) ->
-    emqx_packet:format_truncated_payload(Part, byte_size(Payload), Type).
+    emqx_packet:format_truncated_payload(Part, byte_size(Payload), Type);
+format_payload(Payload, _) ->
+    %% We don't want to crash if there is a field named payload with some other
+    %% type of value
+    Payload.
```

Comment:
If JSON map is the only possible format, maybe we should format it into JSON and truncate.
Otherwise it can be quite bloated when tracing a large SQL output.

Reply:
One will probably use some kind JSON viewer when viewing the JSON traces (for example, the one provided by the dashboard rule tester). This will allow the user tor expand to collapse fields as needed which I think remove the bloat problem that you mention above.

I think the tracing feature is mostly intended to inspect the flow triggered by a few messages or rule applications and not to inspect massive amount of data which makes the bloat problem less relevant. As user using the rule tester in the dashboard I would like to see the whole rule output in a structured way and getting a truncated JSON string (due to some hard coded limit) could probably be quite annoying. So I don't think we should encode JSON maps as text and then truncate them.

Reply:
fair enough.